### ISSUE TITLE
Biostars 3.0 - new features

### ISSUE THREAD

**Issue by ialbert** [ialbert/biostar-central#291]
Features for Biostars 3.0 (scheduled for Spring 2015)

Brain dump of planned features:

- Seamless migration from 2.0 to 3.0
- Reduce codebase at least 10 fold - simplify the templates and views.
- Radically simple installation --> a true ZERO installation codebase where it is possible to run Biostars for a medium site (10 requests per second) on a standard desktop class computer with 1GB RAM and without installing any other software (not even a front end webserver). Everything will be included in a single zip file + executable (app or exe) that can be executed with the default python installation on every platform: Linux, Mac or Windows. (Heavy load deployments will of course need to be customized as before)
- Content Federation and Exchange - biostars sites may interconnect and share user data (login) as well as content either fully or partially (titles only)
- File and image uploads - users may upload images and files and attach them to posts, user reputation determines the size of the uploads
- Simpler account generation - but new content by untrusted users will have to be approved by existing users before being posted (spam protection)
- Event manager - organize conferences/short courses/hackatons with Biostar
- User lists and bookmarking improvements: Users can add their own tags to posts, and organize site content in a structure they think makes best sense (custom lists of posts). 
- Search improvements - sorting by date, user preference, post views
- Content creation customization: provide more choices on the content editor: HTML or Markdown
- Site customization: templating should be possible without editing code in Biostars.
- Data integration: provide web frontend to an existing bittorrent server (opentracker)

**Comment by tvvocold** [comment]
SEO Friendly

**Comment by ialbert** [comment]
@bgruening - I can see the value for it especially when providing support - perhaps in the way social sites do it where adding a username with the @ symbol subscribes the user the the thread. It might need to be limited to moderators to avoid having people added against their will. We'll look at this.

**Comment by satra** [comment]
@ialbert - one possibility is to do it the way github does it. the `@user ` can be used for anybody on github, but any user can unsubscribe from a thread.

**Comment by bgruening** [comment]
@ialbert sounds like a good plan. It can also work like the bitbucket pull-request. Here you can assign peoples from a drop-down list. I would only add this features for admins or people with more than 1k points. Not sure if I would show this to everyone.

**Comment by dannon** [comment]
Adding to @natefoo's request -- a moderation toggle to 'move this item' providing drag and drop functionality to quickly fix comments that were added as answers, etc., would be great.  It's a common problem for us that folks just keep adding answers to posts when they should be replies to comments, replies to other answers, etc.

**Comment by tvvocold** [comment]
check this out! 
 http://segmentfault.com/

**Comment by jennaj** [comment]
Consider adding "News" and "Tutorials" to the top menu bar. It is difficult to search for posts of these types where all are listed. (could be specifically for Galaxy Biostar, we use these quite bit now, and plan to expand Tutorials much more going forward). This would filter posts ideally in the same way that "Forum" does. 

General news can be found under "Planet" but this is just from the RSS feeds (correct?). Not the same as locating posts of type "News: ....".

**Comment by tvvocold** [comment]
I find some sites might be helpful:
1, http://www.telesc.pe/
2, http://segmentfault.com/
3, http://www.v2ex.com/
4, http://www.quora.com/
5, https://news.ycombinator.com/
6, http://www.zhihu.com

**Comment by tvvocold** [comment]
how about add PM ( private message) and "@" feature just like twitter/github.

**Comment by ialbert** [comment]
the site used to have PM messaging but it wasn't that used that much.

The new biostar has a "groups" feature that an be set to either public or
private. This will allow people to discuss issues in a more restricted
manner. More info on groups:

A group operates as a subdomain of the site. It may have a different logo
(and possibly may load extra CSS styling). For example

conference-2015.biostars.org

or

lectures.biostars.org

A group visibility may be public or restricted to members only. In addition
groups may allow members to join freely or members need to be added by the
group owners.

**Comment by kikocorreoso** [comment]
Easy translations. Now it seems all is hardcoded on templates.

**Comment by kikocorreoso** [comment]
Python 3 support.

**Comment by ialbert** [comment]
@kikocorreoso Translations are actually very difficult to get right because once one goes beyong simple word translations and attempts translating sentences with variable content the wording typically changes and leads to more complications.  Biostar 1.0 used to have support for translations via the django internalization framework and had translations for two languages:

https://github.com/ialbert/biostar-central/tree/biostar1/main/server/locale

Back then I put in a lot of work into that yet it was not used. Long story short the feature will need a champion.

Python 3.0 readiness - I have a somewhat negative opinion of Python 3.0 and I am still not convinced it was a good idea or that it will actually gain wide adoption comparable to 2.7. Strangely Python 2.7 offers massive benefits in not actually changing the language hence the system will work reliably right out of the box on many systems. 

The plan is  to Python 3.0 once OS X is shipped with 3.0 as default. In the meantime we'll try to shape the code to be Python 3.0 ready.

**Comment by kikocorreoso** [comment]
@ialbert 

Translations: so the options are to hardcode the templates and  maintain a separate fork with a different templates folder with the translations? :-( It is not a hard work, I already made it but it is not the ideal workflow.

Python 3: I don't want to start a 2vs3 flamewar. That's not mi intention. The scientific python stack ported most of the code years ago. I've been using only Python 3 99% of the time since two years  ago. I maintain some small code repos compatible with Python2 and 3 and there are a lot of big examples with a single codebase for both python versions. Also, Python2 has a scheduled end. Why not start trying to port the code to a single codebase compatible with both python versions?

BTW, thanks for an excellent tool.

P.D.: I'm not trolling, just honest questions for new features :smile:

**Comment by ialbert** [comment]
@kikocorreoso translations don't need a fork just pull requests where the template content that needs to be translated are wrapped into {% trans %} blocks - it is easy to do just needs to be done. On Python 3 I am following some of the porting guidelines to make code easy to covert. It is just that I am very pressed for time and I can't take on the responsibility of testing on both and ensuring that it works on both, I had to pick a side and that was 2.7. I know about dual codebase, a tool I wrote (pyblue) does run on both codebases.

**Comment by ialbert** [comment]
@kikocorreoso just to make sure we are referring to the same concept, translations are typically done via the internationalization framework like so: https://docs.djangoproject.com/en/1.7/topics/i18n/

**Comment by kikocorreoso** [comment]
@ialbert I talk about text translations, not internationalization. 

I saw there are some `trans` or `blocktrans` in the templates (https://github.com/ialbert/biostar-central/blob/master/biostar/server/templates/socialaccount/signup.html#L27) but in the same template there are comments not in these `trans` tags/blocks (https://github.com/ialbert/biostar-central/blob/master/biostar/server/templates/socialaccount/signup.html#L40). I'm talking about to have a full translation of the site with the site available in english, french or spanish depending the community. 

Maybe tools like django-rosetta could help: https://github.com/mbi/django-rosetta

**Comment by ialbert** [comment]
IMHO this should be discussed in a separate issue because it does not
depend on 3.0 and is not a new feature moreover it would get better
exposure on its own.

But just to clarify, translations are a subset of internationalization.
Besides translating the text one may also need to change the formatting of
time, dates, names, numbers, units , change pluralizing etc.  hence the
broader concept of internationalization.

**Comment by biomonika** [comment]
Often users are seeking advice related with Galaxy and don't know about existence of https://biostar.usegalaxy.org/. If the keyword entered is "galaxy", it would be nice if the user was redirected to the Galaxy forum.

**Comment by tvvocold** [comment]
Bugs in biostar v2.0 
Cannot sign up when a spam user exist.(doesn't own the email address but had sign up and cannot confirm 'his' mails)
The user must click the confirm email to activate the account for spam/bad sing up.

**Comment by ialbert** [comment]
Most major features have been coded and functional for 3.0. A public beta will commence in the near future once docs are in better shape.

Python 3 fans will be pleased to know that Biostars now runs on both Python 2 and 3 on the same codebase.

https://travis-ci.org/ialbert/biostar-central/builds/55182711

**Comment by satra** [comment]
great job @ialbert

**Comment by satra** [comment]
ps. @ialbert - let us know when we can:

1. implement it for qa.mit.edu
2. update neurostars

**Comment by tvvocold** [comment]
Hi @ialbert ,
Have you consider mv the doc to [wiki](https://github.com/ialbert/biostar-central/wiki)?

**Comment by ialbert** [comment]
@tvvocold We'll do better than that, the docs can be self hosted by Biostar itself. There will be functionality to push and synchronize local files in a directory with Biostar posts. There will be a command line script, let's call it `biostar` for now

    biostar pull --uuid "my-post","your-post", "someone_else's_post" --dir "somedir"
 
will produce three files in a directory that track the posts with the ids as above. The files have the metadata embedded in them as comments that connect the file to Biostar ids. At any time one can do a sync to get the latests version of these posts more over they can change these posts and then do a `push` of all documents in a directory with 

   biostar push --dir "somedir"

that will look at files with metadata and if changed will go  deposit them back on Biostar  (as long as the user has write access to these posts of course). So for example the Biostar docs  will be markdown documents on GitHub that are also connected to Biostar posts. 

@satra we'll let you know how things progress

**Comment by kikocorreoso** [comment]
@ialbert Excellent news the Python3 compatibility!!!

Thank you very much.

**Comment by tvvocold** [comment]
@ialbert  Cool! 
I got some advice.For now i cannot see the beta verison, so i am going to talk about v2.
1, When user post  a Forum, Job, Blog ...it shoudn't be "answers" in the home page, i think it's better to change it to "comments".
```
<div class="box unanswered">
            <div class="c">*</div>
            <div class="t">answers</div>
        </div>
```

2, Add a Tag/Topic description just like [Quora](https://www.quora.com/Quora)

3, When Tag/Topic goes to 100+, it's better to add a Follow button.

I hope [Quora' FAQ](https://www.quora.com/How-do-I-get-started-using-Quora) is going to be helpful.
Thank you!

**Comment by gregcaporaso** [comment]
@ialbert, was wondering if you guys have an updated ETA on when biostars 3 will go live. I'd like to make it the primary support forum for [An Introduction to Applied Bioinformatics](http://www.readIAB.org) as soon as possible. Thanks!

**Comment by ialbert** [comment]
The codebase is ready - all planned features have been implemented and most (but not all) have exhaustive tests. We lack tests to some of the more elaborate features. 

The real holdback is that I am heading out to vacation in two weeks and I don't want to add the worry of launching right before leaving on a trip.  If we can get most tests done we may launch before July 20, but if not the launch will be in August. There is a hard deadline for August launch as the I want to site live before the semester starts on August 20th.

**Comment by scgruber** [comment]
I've opened #374 to track priorities for test coverage.

**Comment by mdshw5** [comment]
How about the ability to upvote answers on mobile.

**Comment by gregcaporaso** [comment]
Hi all, I wanted to check in on this - is there a target release date for 3.0? Thanks!

**Comment by ialbert** [comment]
what happened is that the coding is finished but I am in the  situation that I don't like how it all turned out. The code is all there should be working, but if I am not sure I achieved what I wanted to do. I am considering various alternatives about what the right course of action should be next.

Forging ahead with 3.0?  backporting the best 3.0 features into 2.0? or doing it again differently jumping 4.0 all on the table

**Comment by ialbert** [comment]
Lately I have been actually "furiously" coding up versions of what I call **biostar4** I am prototyping systems with different paradigms. I went through a semi working prototype coded up in the following systems:

* meteor: https://www.meteor.com/
* flask: http://flask.pocoo.org/
* bottle: http://bottlepy.org/docs/dev/index.html

Finally I think I am converging on something that seems to be very promising. Django front end over a MongoDB backend. It allows me to also reuse a lot of code that I like from biostar3. It radically simplifies storing data. Mongodb is really a document database that seems ideally suited to storing the type of content I have now.

I am trying to functional test version up this weekend to gauge some of its potential and perhaps attract a few collaborators.

**Comment by foreveremain** [comment]
this all sounds really promising !!  I'll wait on more details ...

**Comment by gregcaporaso** [comment]
Ok, thanks. We're still interested in potentially using BioStars as our QIIME (and other software) forum (currently >3000 users in our Google Group-based forum), but we're getting to the point where we need something different soon. Let us know if we can be of help with testing.

**Comment by ialbert** [comment]
Biostar 4.0 early demo, runs off MongoDB

* http://test.biostars.org/
* Track progress here: https://github.com/ialbert/biostar-central/tree/4.0

**Comment by hlj2722** [comment]
If support multiple language function would be great.

**Comment by ialbert** [comment]
@hlj2722 django does support that via internationalization, I have also attempted that in the past like (for Biostar version 1) but it was a lot of work and it will need to have a champion that does the work for the first alternative language:

https://github.com/ialbert/biostar-central/tree/biostar1/main/server/locale

Once the first step is done other languages are easier to add.

**Comment by ialbert** [comment]
Continued here https://github.com/ialbert/biostar-central/issues/400
